Fix Part of #11496: Lint Check for E2E Action Functions

U8NWXD opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR part of #11496.
2. This PR does the following: Adds a lint check to ensure e2e tests use `action.js` functions.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

No user-facing interface. To confirm that the test currently passes, see the CI checks on this PR. I confirmed locally that the test fails when it should.

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

U8NWXD committed:
Add lint check for e2e action scripts

U8NWXD committed:
Add exclude list for end-to-end-action-checks lint test

U8NWXD committed:
Fix end-to-end-action-checks to exclude by base filename

U8NWXD committed:
Shorten action checks rule name

U8NWXD committed:
Merge branch 'develop' into e2e-action-lint

U8NWXD committed:
Fix e2e-action lint failure

U8NWXD commented:
@kevintab95 @DubeySandeep PTAL

DubeySandeep commented on scripts/linters/custom_eslint_checks/rules/e2e-action-exclude.json:
```diff
@@ -0,0 +1,62 @@
+{
+    "exclude": [
+        "ClassroomPage.js",
```
Why are these excluded?

U8NWXD replied:
Because they haven't been fixed yet. They will be removed as we complete tasks in #10798

DubeySandeep commented on scripts/linters/custom_eslint_checks/rules/e2e-action-exclude.json:
```diff
@@ -0,0 +1,62 @@
+{
+    "exclude": [
+        "ClassroomPage.js",
```
We are trying to avoid having oppia project-specific things/string in the linter project. Can you please check whether we can pass these configs through `.eslintrc`?

U8NWXD replied:
done

DubeySandeep reviewed (changes requested):
@U8NWXD, Thanks for the PR, I took a quick pass and left two comments related to the structure of the project, PTAL!

U8NWXD committed:
Limit e2e-action checks to e2e files

U8NWXD committed:
Move e2e-action file include/exclude lists to .eslintrc

U8NWXD committed:
Add files to e2e-action exclude list

U8NWXD commented:
@DubeySandeep PTAL

DubeySandeep commented on .eslintrc:
```diff
@@ -13,14 +13,93 @@
+    {
+      // We only run the e2e action checks on end-to-end test files, and
+      // we exclude files that we haven't fixed yet.
+      "files": [
```
This is great! Thanks for adding these fields here.

DubeySandeep commented on .eslintrc:
```diff
@@ -13,14 +13,93 @@
+      }
     }
-  }],
+
```
Remove unexpected newline.

U8NWXD replied:
done

DubeySandeep commented on scripts/linters/custom_eslint_checks/rules/e2e-action.js:
```diff
@@ -0,0 +1,92 @@
+              elementName: elementName,
+            },
+          });
+        }
```
Maybe have a propertyNameToMessageId dict and have only one if check:
```
if callee.property.name in propertyNameToMessageId:
     context.report({
            node: callee,
            loc: callee.loc,
            messageId: propertyNameToMessageId[callee.property.name],
            data: {
              elementName: elementName,
            },
          });
```

DubeySandeep replied:
Also, is there any issue with having the message key the same as propertyName?

U8NWXD replied:
I've made the message key the same as the property name and simplified the logic as you suggested

DubeySandeep commented on scripts/linters/custom_eslint_checks/rules/e2e-action.js:
```diff
@@ -0,0 +1,92 @@
+        }
+        let elementName = callee.object.name;
+        if (typeof elementName === 'undefined') {
+          elementName = '(some expression)';
```
Can you please explain which situation this if check handles?

U8NWXD replied:
This handles cases where we're calling `.click()`, `.clear()`, etc. on an expression that returns an element rather than an element variable. For example:

```js
element(by.css('.protractor-test-foo')).click()
```

There are examples in the spec file

DubeySandeep replied:
Maybe call this [Element] instead of `(some Expression)`? (not sure!)

U8NWXD replied:
I prefer `(some expression)` because we this is specifically for when we are calling `.click()` on an expression that resolves to an element

DubeySandeep commented on scripts/linters/custom_eslint_checks/rules/e2e-action.js:
```diff
@@ -0,0 +1,92 @@
+    type: 'problem',
+    docs: {
+      description: (
+        'The functions in action.js should be used wherever possible ' +
+        'instead of interacting with elements directly.'),
```
Do you think we should phrase it something around the lines of `All the action on the DOM elements should be performed via the wrapper function defined in action.js`? (not sure)

U8NWXD replied:
I like that the existing text specifies the action we don't want the user to take (interacting with elements directly)

DubeySandeep commented on scripts/linters/custom_eslint_checks/rules/e2e-action.js:
```diff
@@ -0,0 +1,92 @@
+    fixable: null,
+    schema: [],
+    messages: {
+      directElemClick: (
```
Suggestion/optional:

Should we have this check work for all the functions defined in the actions.js file so that if in the future we add new functions in the actions.js file then it works automatically? (Note: We will have to make sure that the function name in action.js matched the element's methods.)

A possible way to do it:

1. Only have one message `'{{elementName}}.{{actionName}}() is called instead of using ' +
        'action.{{actionName}}()'`
2. Fetch the name of all the functions exported by actions.js [User can mention wrapperModulePath as action.js through .eslintrc]
3. In CallExpression function If any callee.property.name matches action's function name then report an issue.

U8NWXD replied:
This works for some action functions like `click()`, but not for others like `select2()`. Therefore, I think we need to explicitly define the criteria for when each action function should be used, not handle it automatically

DubeySandeep commented on scripts/linters/custom_eslint_checks/rules/e2e-action.js:
```diff
@@ -0,0 +1,92 @@
+              elementName: elementName,
+            },
+          });
+        } else if (callee.property.name === 'clear') {
```
Does it covers all the wrapper functions defined in action.js?

U8NWXD replied:
Yes. The `select` and `select2` functions aren't mentioned explicitly, but cases where they should be used will be caught since they will use `.click()`

DubeySandeep reviewed (changes requested):
@U8NWXD, Thanks for the update, it looks great, I've left a few comments and one optional suggestion on restructuring the linter differently, PTAL. (Let me know if you want to discuss the restructure related comment.)

kevintab95 reviewed (approved):
LGTM for codeowner file. Thanks @U8NWXD!

U8NWXD committed:
Merge branch 'develop' into e2e-action-lint

U8NWXD committed:
Remove Lint Check for action.clear

Rich text editors have a .clear() function, so we cannot use a lint
check to find cases where we want to use action.clear instead.

U8NWXD committed:
Remove checks for clear from e2e-action.spec.js

U8NWXD committed:
Use function name as messageId in e2e-action lint check

U8NWXD committed:
Remove mistakenly committed tmp.py file

U8NWXD committed:
Remove unnecessary newline from .eslintrc

U8NWXD committed:
Simplify e2e-action lint logic

U8NWXD commented:
@DubeySandeep I've addressed your comments. PTAL

DubeySandeep commented on core/tests/protractor_utils/CollectionEditorPage.js:
```diff
@@ -159,7 +159,9 @@ var CollectionEditorPage = function() {
     // converted to action.sendKeys(..)
     // because doing so causes the e2e test to fail with the error
     // "Failed: Cannot read property 'bind' of undefined".
```
[Future work] This looks like some general binding issue, instead of using a pragma and TODO, I think we should prioritize fixing this to avoid the further copy of this code.

U8NWXD replied:
fixed

U8NWXD committed:
Replace elem.sendKeys with action.select2

U8NWXD committed:
Remove click before select2
